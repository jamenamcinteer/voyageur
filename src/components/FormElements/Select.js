import React, { useState } from "react";

const Select = props => {
  const style = {
    container: {
      margin: "20px 0"
    },
    label: {
      display: "block",
      fontSize: ".8em",
      fontFamily: "Roboto",
      fontWeight: "bold",
      color: props.theme.darkFont
    },
    input: {
      width: "100%",
      borderRadius: "5px",
      border: `2px solid ${props.theme.themeColorSecondary}`,
      fontSize: "1em",
      padding: "7px 10px",
      color: props.theme.darkFont
    },
    inputFocus: {
      outline: "0",
      border: `2px solid ${props.theme.themeColor}`
    }
  };

  const [inputStyles, setInputStyles] = useState(style.input);

  const ID =
    "_" +
    Math.random()
      .toString(36)
      .substr(2, 9);

  const [inputValue, setInputValue] = useState(props.value ? props.value : "");

  return (
    <div style={style.container}>
      <label style={style.label} htmlFor={ID}>
        {props.label}
      </label>
      <select
        style={inputStyles}
        onFocus={e => setInputStyles({ ...style.input, ...style.inputFocus })}
        onBlur={e => setInputStyles(style.input)}
        onChange={e => {
          setInputValue(e.target.value);
          props.handleChange(e.target.value);
        }}
        value={inputValue}
        id={ID}
      >
        {props.placeholder && <option value="">{props.placeholder}</option>}
        {props.options.length > 0 &&
          props.options.map(option => {
            return (
              <option key={option.value} value={option.value}>
                {option.text}
              </option>
            );
          })}
      </select>
    </div>
  );
};

export default Select;
