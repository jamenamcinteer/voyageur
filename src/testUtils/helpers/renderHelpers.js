import React from "react";
import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import theme from "../../theme";
import "jest-styled-components";
import { createStore } from "redux";
import { Provider } from "react-redux";
import tripsReducer, { tripsReducerDefaultState } from "../../reducers/trips";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";

export function renderWithTheme(component) {
  return render(<ThemeProvider theme={theme}>{component}</ThemeProvider>);
}

export function renderWithReduxRouterAndTheme(
  ui,
  {
    initialState,
    store = createStore(tripsReducer, tripsReducerDefaultState),
    route = "/",
    history = createMemoryHistory({ initialEntries: [route] })
  } = {}
) {
  return {
    ...render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <Router history={history}>{ui}</Router>
        </ThemeProvider>
      </Provider>
    ),
    store,
    history
  };
}
