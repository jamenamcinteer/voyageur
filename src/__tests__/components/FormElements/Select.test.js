import React from "react";
import { cleanup, fireEvent } from "@testing-library/react";
import Select from "../../../components/FormElements/Select";

import { renderWithTheme } from "../../../testUtils/helpers/renderHelpers";

afterEach(cleanup);

jest.mock("uniqid", () => {
  let value = 1;
  return () => value++;
});

const options = [
  {
    value: "1",
    text: "Option 1"
  },
  {
    value: "2",
    text: "Option 2"
  },
  {
    value: "3",
    text: "Option 3"
  }
];
const handleChange = jest.fn();
const label = "Select Options";

test("should render with theme and passed props", () => {
  const { getByLabelText, asFragment } = renderWithTheme(
    <Select options={options} label={label} handleChange={handleChange} />
  );
  expect(asFragment()).toMatchSnapshot();

  const select = getByLabelText(label);

  expect(select.value).toBe(options[0].value);
});

test("should render with passed default value", () => {
  const { getByLabelText } = renderWithTheme(
    <Select
      options={options}
      label={label}
      value={options[2].value}
      handleChange={handleChange}
    />
  );

  const select = getByLabelText(label);

  expect(select.value).toBe(options[2].value);
});

test("should change select value when option is changed", () => {
  const { getByLabelText } = renderWithTheme(
    <Select options={options} label={label} handleChange={handleChange} />
  );

  const select = getByLabelText(label);

  fireEvent.change(select, {
    target: {
      value: options[1].value
    }
  });

  expect(select.value).toBe(options[1].value);
  expect(handleChange).toHaveBeenCalledWith(options[1].value);
  expect(handleChange).toHaveBeenCalledTimes(1);
});
