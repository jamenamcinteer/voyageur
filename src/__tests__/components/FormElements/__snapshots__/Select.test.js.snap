// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render with theme and passed props 1`] = `
<DocumentFragment>
  .c0 {
  margin: 20px 0;
}

.c1 {
  display: block;
  font-size: 0.8em;
  font-weight: bold;
}

.c2 {
  width: 100%;
  border-radius: 5px;
  border: 2px solid #c7d2db;
  font-size: 1em;
  padding: 7px 10px;
  background: #fff;
}

.c2:focus {
  outline: 0;
  border: 2px solid #0C7C88;
}

<div
    class="c0"
  >
    <label
      class="c1"
      for="1"
    >
      Select Options
    </label>
    <select
      class="c2"
      id="1"
    >
      <option
        value="1"
      >
        Option 1
      </option>
      <option
        value="2"
      >
        Option 2
      </option>
      <option
        value="3"
      >
        Option 3
      </option>
    </select>
  </div>
</DocumentFragment>
`;
